PC/launcher2.c

```diff
@@ -491,62 +491,39 @@ dumpSearchInfo(SearchInfo *search)
 
 
 int
-findArgumentLength(const wchar_t *buffer, int bufferLength)
+findArgv0Length(const wchar_t *buffer, int bufferLength)
 {
-    if (bufferLength < 0) {
-        bufferLength = (int)wcsnlen_s(buffer, MAXLEN);
-    }
-    if (bufferLength == 0) {
-        return 0;
-    }
-    const wchar_t *end;
-    int i;
-
-    if (buffer[0] != L'"') {
-        end = wcschr(buffer, L' ');
-        if (!end) {
-            return bufferLength;
-        }
-        i = (int)(end - buffer);
-        return i < bufferLength ? i : bufferLength;
-    }
-
-    i = 0;
-    while (i < bufferLength) {
-        end = wcschr(&buffer[i + 1], L'"');
-        if (!end) {
-            return bufferLength;
-        }
-
-        i = (int)(end - buffer);
-        if (i >= bufferLength) {
-            return bufferLength;
-        }
-
-        int j = i;
-        while (j > 1 && buffer[--j] == L'\\') {
-            if (j > 0 && buffer[--j] == L'\\') {
-                // Even number, so back up and keep counting
-            } else {
-                // Odd number, so it's escaped and we want to keep searching
-                continue;
+    // Note: this implements semantics that are only valid for argv0.
+    // Specifically, there is no escaping of quotes, and quotes within
+    // the argument have no effect. A quoted argv0 must start and end
+    // with a double quote character; otherwise, it ends at the first
+    // ' ' or '\t'.
```

Comment:
Technically speaking, any part of argv0 may be quoted. E.g. `C:\"path to"\"py.exe"` is parsed as `C:\path to\py.exe` by the crt. Being quoted just prevents tab or space from terminating the argument.

Reply:
Please see the associated bug (which I just fixed, because I put the wrong ID in the title originally).

We're following CreateProcess rules here, because it's going to be passed to CreateProcess.

Reply:
@ChrisDenton, The C runtime's more general parsing of `argv[0]` (i.e. multiple quoted parts terminated by space or tab) is only relevant if the executable path is passed explicitly to `CreateProcessW()` in `lpApplicationName`. If `CreateProcessW()` is called with `lpApplicationName` as `NULL`, as the launcher does, then the API parses the command to run from `lpCommandLine`. If the command line begins with a double quote, the API consumes all characters up to the next double quote or the end of the command line. If the command line doesn't begin with a double quote, the API tokenizes on space and tab characters, repeatedly searching until it finds a non-directory file. 

For example, if the command line is `r'C:\Program Files\Python311\python.exe'`, the API will execute "C:\\Program.exe" if it exists. If the command line is `r'"C:\Program Files\Python311\python.exe"spam'`, the API will execute "C:\\Program Files\\Python311\\python.exe" if it exists. Note that there's an inconsistency with the C runtime's quoting rules. The CRT always splits arguments on space or tab. For example:

```pycon
>>> script = 'import sys; print(sys.orig_argv)'
>>> subprocess.call(fr'"{sys.executable}"spam -c "{script}"')
['C:\\Program Files\\Python311\\python.exespam', '-c', 'import sys; print(sys.orig_argv)']
0
```
The API executed "python.exe", not "python.exespam".

Reply:
Yes, that inconsistency is what worried me. As far as I understand it, no normalization is being performed in the python code so the inconsistency remains? So the CRT and `CreateProcessW` will be understanding the arguments differently?